New: Implement cache in order to only operate on changed files since previous run. (fixes #2998)

royriojas opened the pull request:
This is the first draft of adding a cache to Eslint.

- The basic detection of a changed file is deferred to file-entry-cache.
- The file cache is stored in the current working directory (`process.cwd()`) where the command is executed. The file created is called `.eslintcache`
- The cache only stores files that successfully passed validation. The ones that failed are removed to guarantee they will be processed the next run.

an easy way to test is run this in the project

```bash
./bin/eslint.js lib/ --cache # this will take 3.5s aprox

./bin/eslint.js lib/ --cache # this will take 150ms aprox. Obviously because the results are cached

# touch a couple of files, introduce some errors on purpose
./bin/eslint.js lib/ --cache # this will take 200ms aprox. 
```

michaelficarra commented on lib/cli-engine.js:
```diff
@@ -402,6 +457,12 @@ CLIEngine.prototype = {
+            fCache.reconcile();
+        }
+
+        debug("Time: " + (Date.now() - startTime) + "ms");
```
I don't like these implicit conversions. Use `getTime` or unary `+` to explicitly convert dates to numbers.

royriojas replied:
Hi @michaelficarra 

Don't get what the issue is `Date.now()` returns a number (same as `(new Date()).getTime()` )

michaelficarra replied:
My bad, you're right. I was thinking of `new Date`.

michaelficarra commented on lib/cli-engine.js:
```diff
@@ -273,6 +273,17 @@ function CLIEngine(options) {
     this.options = assign(Object.create(defaultOptions), options || {});
 
+    /**
+	 * fileCache used to not operate on files that haven't changed
```
Indentation.

michaelficarra commented on lib/cli-engine.js:
```diff
@@ -273,6 +273,17 @@ function CLIEngine(options) {
+     * since last successful execution
+     * @type {Object}
+     */
+    this.fCache = require("file-entry-cache").create(path.resolve(".eslintcache"));
```
Is there no way to configure this?

royriojas replied:
So you're saying not creating the cache file all the time? Since we need to destroy the cache if the call was not done using `--cache`, it was easier to delegate the deletion of the file to the same module instead of including `rimraf` as a dependency. if that is not a problem, them I can create the cache only if the `--cache` flag is set

michaelficarra replied:
I meant the file name/location. Is there any way that can be configured?

royriojas replied:
yes it can be configured, but @nzakas suggested to use the current working directory, making it a parameter is rather trivial. @nzakas what do you think?

lo1tuma commented:
Looks like `file-entry-cache` relies on `mtime` which could be a problem, because `git checkout` always updates `mtime` even if the file wasn’t changed.

platinumazure commented:
@lo1tuma It wouldn't be any worse than having no cache, right? `git checkout` implies a fresh run is needed, sure, but future runs of `eslint --cache` on the same branch will still get some wins out of this.

royriojas commented:
@lo1tuma Yeah I agree with @platinumazure 

It won't be worse than no cache, I have being using this in [eslinter](https://www.npmjs.com/package/eslinter) for a while with no major issues

gyandeeps commented on tests/lib/cli-engine.js:
```diff
@@ -844,6 +844,133 @@ describe("CLIEngine", function() {
+            var rimraf = require("rimraf").sync;
+
+            beforeEach(function() {
+                rimraf(path.resolve(".eslintcache"));
```
why not just use `fs` module to do this and remove `rimraf` dependency. We have used fs in lot of other tests also.

royriojas replied:
True

royriojas replied:
I will use fs directly

lo1tuma commented:
@royriojas @platinumazure This is a new feature and not a bugfix, so it should have some kind of value that is worth to be added. Anyway, that’s not what I want to discuss. I want to propose that we should find a caching solution which doesn’t rely on `mtime`.

platinumazure commented:
@lo1tuma I work at a shop where we have hundreds, possibly thousands, of JavaScript files in our solution, and linting all of them takes 3 minutes (and that's with only a subset of rules even being on, since we are still transitioning away from JSHint). Being able to lint only some of them as needed would save a lot of development time and encourage people to actually use the linting tools instead of just waiting until right before committing (and inevitably forgetting to do so). So something like this would vastly improve code quality where I work, and that value should speak for itself.

I don't see why relying on mtime would be a problem. When checking out a branch, the entire directory could well change and it absolutely makes sense that eslint should be rerun on all files. If you don't want to use mtime as the indication of an expired cache item, what do you want to use instead?

lo1tuma commented:
@platinumazure I’m not arguing against the caching feature, I’m arguing against using `mtime` for caching. I think a better approach would be to rely on a hash of the file content, similar to the [Tarsum Checksum Version1](https://github.com/docker/docker/blob/e137f2d08187550393048295d8238dc18fe2e06f/pkg/tarsum/tarsum_spec.md) which is used by docker.

royriojas commented:
@lo1tuma @platinumazure I do initially planned to calculate the checksum for each file, but I didn't want to impact the normal linting process, calculating the checksum will have some extra processing time involved. the `mtime` and `size` are kinda cheap in comparison. It can be done a parameter if required

nzakas commented:
Since this a new feature, I think we should be focusing on getting something that people can try out and give us feedback on. Relying on `mtime` *might* be a problem for some people, but we really can't be sure until we put it out there. So, I think it's fine as a first approach, and the feedback will tell us if it's a problem or not.

nzakas commented on lib/cli-engine.js:
```diff
@@ -273,6 +273,17 @@ function CLIEngine(options) {
+     * execution (e.g. file passed with no errors and no warnings
+     * @type {Object}
+     */
+    this.fCache = require("file-entry-cache").create(path.resolve(".eslintcache"));
```
Can you move this `require` up with the others? (Part of our conventions.)

nzakas replied:
Also, I'd just call this `fileCache` or `cache`...the `f` isn't very useful here. :)

nzakas replied:
And we need to consider if this should be private (begin with `_`). Do we want to allow consumers of `CLIEngine` to alter this?

nzakas commented on lib/cli-engine.js:
```diff
@@ -377,18 +388,63 @@ CLIEngine.prototype = {
+            // if we need to use the cache
+            if (options.useCache) {
+                // get the descriptor for this file
+                var fDescriptor = fCache.getFileDescriptor(filename);
```
`fileDescriptor` or `descriptor`

nzakas commented on lib/cli-engine.js:
```diff
@@ -377,18 +388,63 @@ CLIEngine.prototype = {
+
+                // if the file has not changed (either size or mtime is different from previous run)
+                if (!fDescriptor.changed) {
+                    debug("Skipping file since haven't changed: " + filename);
```
hasn't

platinumazure replied:
Since "it hasn't changed", maybe?

nzakas commented on lib/cli-engine.js:
```diff
@@ -377,18 +388,63 @@ CLIEngine.prototype = {
+                // store the file in the cache so we can guarantee that
+                // next execution will also operate on this file
+                if ( res.errorCount > 0 || res.warningCount > 0 ) {
+                    debug("Do not remember files with errors or warnings: " + filename);
```
Maybe something more like "File has problems, skipping:".

nzakas commented on lib/options.js:
```diff
@@ -123,5 +123,11 @@ module.exports = optionator({
+        option: "cache",
+        type: "Boolean",
+        default: "false",
+        description: "Use a cache to avoid process files that haven't changed since last run. The cache is saved in the file .eslintcache in the directory where the command is executed"
```
The description needs to be much shorter, otherwise it wraps like crazy in the command line. Something like, "Only check changed files." will do.

nzakas commented on tests/lib/cli-engine.js:
```diff
@@ -844,6 +844,141 @@ describe("CLIEngine", function() {
+                sandbox.restore();
+                delCache();
+            });
+            it("should remember the files from a previous run and do not operate on then if not changed", function() {
```
then -> them

nzakas commented:
Looks like a good start, just a few nitpicks conventions-wise. Also, can you please update the documentation for the CLI with information about this flag?

royriojas commented:
@nzakas sure, where is that documentation?

platinumazure commented:
@royriojas Look in `docs/user-guide/command-line-interface.md`.

royriojas commented:
@platinumazure thank you :)

royriojas commented:
Hi guys, 

I guess I have addressed all the issues you mention. Also I found some issues for the build in `OSX`. The issues were happening on both `master` and my branch. It seems that the temporary folder in mac returns `/var/something` and the real path is `/private/var/something` don't know if that is the real reason of the failure but I couldn't reproduce that on Linux.

https://gist.github.com/royriojas/9b7fdbbea4c7e772e5a3

Cheers!

gyandeeps commented:
Can ypou please update the `defaultOptions` variable inside `cli-engine` with cache set to false.
And also update the nodejs api docs for `CLIEngine` here: https://github.com/eslint/eslint/blob/master/docs/developer-guide/nodejs-api.md#cliengine

royriojas commented:
hi @gyandeeps,

to test which scenario? that cache if by default false?

nzakas commented on tests/lib/cli-engine.js:
```diff
@@ -927,6 +927,141 @@ describe("CLIEngine", function() {
+
+                var file = getFixturePath("cache/src", "test-file.js");
+
+                engine.executeOnFiles([file]);
```
Can you also check that you get the correct results returned?

royriojas replied:
@nzakas, like the right report?

nzakas replied:
Yes. `executeFiles()` should return the same results regardless of cache being on or off.

royriojas replied:
cool, will add the check for it

royriojas commented:
Hi guys

I have added all changes requested. 

- Added cacheFile as a parameter to set a custom file to save the cache
- Fixed the unit tests to pass when executed on OSX. (`/var/temp/`, was resolved as `/private/var/temp` which break a lot of unit tests in OSX)
- Added unit test to verify the results from an execution without the cache was the same as an execution with the cache
- As a way to test the benefits of the cache, I modified the makefile to use the cache for the `lint` task. This actually reduced the time of the process, because only the modified files were processed. I did that only for local executions on travis it will process all the files, just to keep some sort of peace of mind.

@nzakas, @gyandeeps please let me know.

nzakas commented on Makefile.js:
```diff
@@ -390,10 +390,15 @@ target.all = function() {
     var errors = 0,
-        lastReturn;
+        lastReturn,
+        MAKEFILE_CACHE,
```
Constants should be assigned values immediately.

royriojas replied:
should I declare them with const? or var is ok?

nzakas commented on Makefile.js:
```diff
@@ -390,10 +390,15 @@ target.all = function() {
+        JS_CACHE,
+        TEST_CACHE;
+
+    MAKEFILE_CACHE = process.env.TRAVIS ? " " : " --cache --cache-file .cache/makefile_cache ";
```
Why not use `.eslintcache`?

royriojas replied:
cache is kept only between executions with the same set of files, this is to make sure the cache won't remember files that were deleted or renamed.

Since the linting process in the makefile happens in 3 fases, we need 3 different caches for them

nzakas commented on docs/user-guide/command-line-interface.md:
```diff
@@ -48,6 +48,7 @@ Options:
   --stdin-filename            Specify filename to process STDIN as
   --init                      Run config initialization wizard
   --parser                    Specify the parser to be used
+  --cache                     Only check changed files - default: false
```
Don't forget to add `--cache-file` here.

royriojas replied:
good catch!

nzakas commented on lib/cli-engine.js:
```diff
@@ -302,6 +305,17 @@ function CLIEngine(options) {
+     * execution (e.g. file passed with no errors and no warnings
+     * @type {Object}
+     */
+    this.fileCache = fileEntryCache.create(path.resolve(this.options.cacheFile));
```
Just want to double-check: is there any reason a user of `CLIEngine` would want to access the cache? If not, it should be prepended with an `_` to indicate it's private.

royriojas replied:
I tried, but it failed the linting process :)

nzakas replied:
You can disable the rule in this file of necessary. I'm more concerned with people directly touching an object that they shouldn't. If this should be private, feel free to use the underscore and disable the rule.

nzakas commented on lib/cli-engine.js:
```diff
@@ -453,6 +512,13 @@ CLIEngine.prototype = {
+        }
+
+        debug("Time: " + (Date.now() - startTime) + "ms");
+
```
I think this is the total amount of time it took to lint? If so, maybe "Linting complete in: "

nzakas commented:
Looking good, just a few things.

Also, why the changes for `fs.realpathSync()`?

royriojas commented:
@nzakas to fix this 

https://gist.github.com/royriojas/9b7fdbbea4c7e772e5a3

So in OSX the build fails without `fs.realPathSync`

royriojas commented:
@nzakas done! 

- made _fileCache private by convention
- use variables instead of constants since the value change depending on the environment
- added missing entry to cli interface docs

Regards!

nzakas commented:
This looks great. We just rearranged our CLI options and documentation in #3613. Can you please [rebase](http://eslint.org/docs/developer-guide/contributing/pull-requests#step-3-rebase-onto-upstream) and update this?

Thanks so much for your hard work and patience, this is going to be great!

royriojas commented:
@nzakas. Sure will do the rebase now

royriojas commented:
@nzakas done!

gyandeeps commented:
LGTM.
Great work @royriojas and thanks. :tada:
